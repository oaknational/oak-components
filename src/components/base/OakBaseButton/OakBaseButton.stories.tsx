import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { OakBaseButton } from "./OakBaseButton";

/**
 *
 * An unstyled button to be used as a basis for all UI button components.
 * The following callbacks are available for tracking focus events:
 *
 *
 */

const meta: Meta<typeof OakBaseButton> = {
  component: OakBaseButton,
  tags: ["autodocs"],
  title: "components/base/OakBaseButton",
  argTypes: {},
  parameters: {
    controls: {
      include: [],
    },
  },
};
export default meta;

type Story = StoryObj<typeof OakBaseButton>;

export const Default: Story = {
  render: (args) => <OakBaseButton {...args} />,
  args: {},
  parameters: {
    controls: { include: [] },
  },
};

Default.args = {
  $ba: "border-solid-s",
  $pa: "inner-padding-s",
  $borderRadius: "border-radius-m",
};
