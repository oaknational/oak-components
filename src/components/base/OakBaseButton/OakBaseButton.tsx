import React, { useRef } from "react";
import styled from "styled-components";

import { colorStyle, ColorStyleProps } from "@/styles/utils/colorStyle";
import { displayStyle, DisplayStyleProps } from "@/styles/utils/displayStyle";
import { spacingStyle, SpacingStyleProps } from "@/styles/utils/spacingStyle";
import {
  typographyStyle,
  TypographyStyleProps,
} from "@/styles/utils/typographyStyle";
import {
  dropShadowStyle,
  DropShadowStyleProps,
} from "@/styles/utils/dropShadowStyle";
import { borderStyle, BorderStyleProps } from "@/styles/utils/borderStyle";

type StyledButtonProps = TypographyStyleProps &
  SpacingStyleProps &
  ColorStyleProps &
  DisplayStyleProps &
  BorderStyleProps &
  DropShadowStyleProps;

const StyledButton = styled.button<StyledButtonProps>`
  background: none;
  color: inherit;
  border: none;
  padding: 0;
  font: inherit;
  cursor: pointer;
  text-align: left;
  font-family: unset;
  ${typographyStyle}
  ${colorStyle}
  ${spacingStyle}
  ${displayStyle}
  ${borderStyle}
  ${dropShadowStyle}
`;

export type OakBaseButtonProps = StyledButtonProps & {
  onClick?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
  onHovered?: (
    event: React.MouseEvent<HTMLButtonElement, MouseEvent>,
    duration: number,
  ) => void;
  children?: React.ReactNode;
};

export const OakBaseButton = (props: OakBaseButtonProps) => {
  const { onClick, onHovered, ...rest } = props;

  const hoverStart = useRef(Date.now());

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    if (onClick) {
      onClick(event);
    }
  };

  const handleMouseEnter = () => {
    hoverStart.current = Date.now();
    console.log("handleMouseEnter", hoverStart.current);
  };

  const handleMouseLeave = (event: React.MouseEvent<HTMLButtonElement>) => {
    const delta = Date.now() - hoverStart.current;
    if (onHovered) {
      onHovered(event, delta);
    }
  };

  return (
    <StyledButton
      {...rest}
      onClick={handleClick}
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
    >
      {props.children}
    </StyledButton>
  );
};
