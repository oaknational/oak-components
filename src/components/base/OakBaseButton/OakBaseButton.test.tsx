import React from "react";
import "@testing-library/jest-dom";
import { render, fireEvent } from "@testing-library/react";
import { create } from "react-test-renderer";

import { OakBaseButton } from "./OakBaseButton";

describe("OakBaseButton", () => {
  afterEach(() => {
    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });

  it("renders", () => {
    const { getByTestId } = render(
      <OakBaseButton data-testid="test">Click</OakBaseButton>,
    );
    expect(getByTestId("test")).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const tree = create(<OakBaseButton>Click Me</OakBaseButton>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders the chidren", () => {
    const { getByText } = render(<OakBaseButton>Click</OakBaseButton>);
    expect(getByText("Click")).toBeInTheDocument();
  });

  it("calls onClick method", () => {
    const onClick = jest.fn();
    const { getByTestId } = render(
      <OakBaseButton data-testid="test" onClick={onClick}>
        Click
      </OakBaseButton>,
    );
    getByTestId("test").click();
    expect(onClick).toHaveBeenCalled();
  });

  it("calls onHovered method when a mouseover and mouseout event has happened", () => {
    const onHovered = jest.fn();
    const { getByTestId } = render(
      <OakBaseButton data-testid="test" onHovered={onHovered}>
        Click
      </OakBaseButton>,
    );
    fireEvent.mouseEnter(getByTestId("test"));
    fireEvent.mouseLeave(getByTestId("test"));
    expect(onHovered).toHaveBeenCalledTimes(1);
  });

  it("calls doesn't call onHovered method before a mouseout event happens", () => {
    const onHovered = jest.fn();
    const { getByTestId } = render(
      <OakBaseButton data-testid="test" onHovered={onHovered}>
        Click
      </OakBaseButton>,
    );
    fireEvent.mouseEnter(getByTestId("test"));
    expect(onHovered).not.toHaveBeenCalled();
    fireEvent.mouseLeave(getByTestId("test"));
    expect(onHovered).toHaveBeenCalledTimes(1);
  });

  it("correctly captures the duration of the hover event", () => {
    jest.useFakeTimers();
    const onHovered = jest.fn();
    const { getByTestId } = render(
      <OakBaseButton data-testid="test" onHovered={onHovered}>
        Click
      </OakBaseButton>,
    );
    fireEvent.mouseEnter(getByTestId("test"));
    jest.advanceTimersByTime(1000);
    fireEvent.mouseLeave(getByTestId("test"));
    expect(onHovered).toHaveBeenCalledWith(expect.anything(), 1000);
  });
});
