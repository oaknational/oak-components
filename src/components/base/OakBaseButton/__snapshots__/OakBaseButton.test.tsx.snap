// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OakBaseButton matches snapshot 1`] = `
<button
  className="sc-aXZVg jxlkHt"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  Click Me
</button>
`;
